src/tools/ip/quicip.h

```diff
@@ -110,7 +114,11 @@ ClientConnectionCallback(
         }
         break;
     case QUIC_CONNECTION_EVENT_SHUTDOWN_COMPLETE:
-        Context->MsQuic->ConnectionClose(Connection);
+        {
+            std::scoped_lock Lock{Context->DoneMutex};
```

Comment:
Not sure what the difference is, but in spinquic.cpp we use `std::lock_guard` instead. Perhaps we should use the same to be consistent in our code base?

Reply:
Actually, I don't quite understand the need for a lock at all. All you should need to do is have the `DoneEvent` and wait on it forever. Keep the connection clean up logic in SHUTDOWN_COMPLETE. No need to explicitly call shutdown. No need for `IsDone`. We already set the idle timeout to 2 seconds, so that will kill the connection if it gets stuck.

Reply:
You can't wait on a condition_variable in C++ without a mutex. Its a requirement for the API. And you need IsDone otherwise you have a race condition where the connection closes before you actually start the wait. I'll move the close back however.